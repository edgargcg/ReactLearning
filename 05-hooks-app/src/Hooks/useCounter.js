import { useState } from 'react';

export const useCounter = ( initialValue = 10 ) => {
    const [counter, setCounter] = useState(initialValue)

    const increment = ( factor = 1 ) => {
        setCounter( counter + factor );
    }

    const decrement = ( factor = 1 ) => {
        setCounter( counter - factor );
    }

    const reset = ( resetValue = initialValue ) => {
        setCounter(resetValue);
    }

    return {
        counter,
        reset,
        decrement,
        increment,
    };
}
