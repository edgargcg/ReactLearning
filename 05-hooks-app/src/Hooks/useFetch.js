import { useEffect, useRef, useState } from "react"

export const useFetch = ( url ) => {
    const [state, setState] = useState({
        data: null,
        error: null,
        loading: true
    });

    const isMounted = useRef(true);

    useEffect(() => {
        return () => {
            isMounted.current = false;
        }
    }, [])

    useEffect(() => {
        
        setState(
            {
                data: null,
                error: null,
                loading: true
            }
        );

        fetch( url )
            .then( resp => {
                if (!resp.ok){
                    throw new Error('Could not fetch data from resource');
                }

                return resp.json()
            }) 
            .then( data => {

                if (isMounted){
                    setState({
                        data: data,
                        error: null,
                        loading: false,
                    });
                }

            }).catch(error => {
                if (isMounted){
                    setState({
                        data: null,
                        error: error,
                        loading: false,
                    });
                }
            });

    }, [url]);

    return state;
}
