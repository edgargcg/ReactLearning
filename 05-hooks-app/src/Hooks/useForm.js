import { useState } from 'react'

export const useForm = ( initialState = {} ) => {

    const [values, setValues] = useState(initialState);

    const handleInputChange = (e) => {
        const { name, value } = e.target;

        setValues(
            {
                ...values,
                [name]: value
            }
        );
    }

    return[
        values,
        handleInputChange
    ];

}
