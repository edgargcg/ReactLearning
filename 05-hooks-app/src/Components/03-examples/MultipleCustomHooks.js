import React from 'react';
import { useFetch } from '../../Hooks/useFetch';
import { useCounter } from '../../Hooks/useCounter';

import './MultipleCustomHooks.css';

export const MultipleCustomHooks = () => {
    const { counter, increment } = useCounter(1);
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    const { author, quote } = !!data && data[0];

    return (
        <div>
            <h1>Breaking Bad Quotes</h1>
            <hr />
            {
                loading 
                    ?
                        (
                            <div className="alert alert-info text-center">
                                Loading..
                            </div>
                        )
                    :
                        (
                            <blockquote className="blockquote text-right">
                                <p>
                                    { quote }
                                </p>
                                <footer className="blockquote-footer">
                                    { author }
                                </footer>
                            </blockquote>
                        )
            }
            <button 
                className="btn btn-primary"
                onClick={ () => increment(1) }
            >
                Next quote
            </button>
        </div>
    )
}
