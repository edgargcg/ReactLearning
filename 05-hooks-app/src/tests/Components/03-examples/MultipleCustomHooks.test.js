import { shallow } from 'enzyme'
import { MultipleCustomHooks } from '../../../Components/03-examples/MultipleCustomHooks'
import { useFetch } from '../../../Hooks/useFetch';

jest.mock('../../../Hooks/useFetch');

describe('Test on <MultipleCustomHooks />', () => {
    
    test('should render <MultipleCustomHooks />', () => {
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        
        expect( wrapper ).toMatchSnapshot();
    })
    
    test('should display information', () => {
        useFetch.mockReturnValue({
            data: [{
                author: 'Edgar Campos',
                quote: 'Hola Mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);
    
        expect( wrapper.find('.alert').exists() ).toBe( false );
        expect( wrapper.find('.mb-0').text().trim() ).toBe( 'Hola Mundo' );
        
    })
    
    

})
