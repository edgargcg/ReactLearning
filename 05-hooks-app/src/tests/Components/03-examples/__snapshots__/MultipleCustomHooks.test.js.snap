// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <MultipleCustomHooks /> should render <MultipleCustomHooks /> 1`] = `
<div>
  <h1>
    Breaking Bad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading..
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
